### ISSUE TITLE
[Rule Proposal] Enforce (not) using a labelled return in (multiline) lambda bodies

### ISSUE THREAD

**Issue by FWest98** [pinterest/ktlint#3207]
## Expected Rule behavior
The Kotlin Coding Conventions [state](https://kotlinlang.org/docs/coding-conventions.html#returns-in-a-lambda):
> Do not use a labeled return for the last statement in a lambda.

It would be nice to have a rule to check this. In addition, if possible, it would be nice to have a check for the opposite behaviour for multiline bodies specifically - this is the preferred style in our team but I have seen this in more places. Neither of these conflicts with IntelliJ formatting, since that does not touch labelled returns.

Concretely, I would expect the following behaviour:
```kt
// Accepted
fun foo() {
    val x = lambda {
        "value"
    }
}

// Rejected
fun foo() {
    val x = lambda {
        return@lambda "value"
    }
}

// Accepted when multiline bodies do not require a labelled return (default)
fun foo() {
    val x = lambda {
        doSomething()
        "value"
    }
}

// Accepted when multiline bodies do require a labelled return (configurable)
fun foo() {
    val x = lambda {
        doSomething()
        return@lambda "value"
    }
}
```

**Comment by paul-dingemans** [comment]
Tnx for this proposal. It sounds like a good idea as it is an official recommendation.

**Comment by paul-dingemans** [comment]
> Do not use a labeled return for the last statement in a lambda.

It turned out that this is too generic. A labeled return which is the last statement in a lamba can only (safely) be removed when the label refers to that lambda itself. If for some reason the labeled return refers to an outer lambda, it may not be removed as it can lead to compilation issue, or changed behavior. See unit tests below for my understanding of nested lambda's and labeled returns. It is fixed in #3358

```
@Nested
inner class ReturnInLambda {
    @Test
    fun `Return in lambda works as expected, the result is not mapped by the outer mapping`() {
        fun someFunction(): Result<String> =
            run outer@{
                run {
                    runCatching { /* All goes well */ }
                        .map { return Result.success("It was a success") }
                        .map { "Are you sure?" }
                }.map { "$it (Remapped by outer)" }
            }.map { "final result: $it" }

        assertThat(someFunction().getOrNull()).isEqualTo("It was a success")
    }

    @Test
    fun `Return @outer in lambda works as expected, the result is not mapped by the outer mapping, but is wrapped by mapping outside outer`() {
        fun someFunction(): Result<String> =
            run outer@{
                run {
                    runCatching { /* All goes well */ }
                        .map { return@outer Result.success("It was a success") }
                        .map { "Are you sure?" }
                }.map { "$it (Remapped by outer)" }
            }.map { "final result: $it" }
        assertThat(someFunction().getOrNull()).isEqualTo("final result: It was a success")
    }

    @Test
    fun `Return @run in lambda works as expected and the result is mapped by the outer mapping`() {
        fun someFunction(): Result<String> =
            run outer@{
                run {
                    runCatching { /* All goes well */ }
                        .map { return@run Result.success("It was a success") }
                        .map { "Are you sure?" }
                }.map { "$it (Remapped by outer)" }
            }.map { "final result: $it" }
        assertThat(someFunction().getOrNull()).isEqualTo("final result: It was a success (Remapped by outer)")
    }

    @Test
    fun `Return @map in lambda does not work as expected`() {
        fun someFunction(): Result<String> =
            run outer@{
                run {
                    runCatching { /* All goes well */ }
                        .map { return@map Result.success("It was a success") }
                        .map { "Are you sure?" }
                }.map { "$it (Remapped by outer)" }
            }.map { "final result: $it" }
        assertThat(someFunction().getOrNull()).isEqualTo("final result: Are you sure? (Remapped by outer)")
    }

    @Test
    fun `Without return in lambda does not work as expected`() {
        fun someFunction(): Result<String> =
            run outer@{
                run {
                    runCatching { /* All goes well */ }
                        .map { Result.success("It was a success") }
                        .map { "Are you sure?" }
                }.map { "$it (Remapped by outer)" }
            }.map { "final result: $it" }
        assertThat(someFunction().getOrNull()).isEqualTo("final result: Are you sure? (Remapped by outer)")
    }
}
```

        @Test
        fun `Without return in lambda does not work as expected`() {
            fun someFunction(): Result<String> =
                run outer@{
                    run {
                        runCatching { /* All goes well */ }
                            .map { Result.success("It was a success") }
                            .map { "Are you sure?" }
                    }.map { "$it (Remapped by outer)" }
                }.map { "final result: $it" }
            assertThat(someFunction().getOrNull()).isEqualTo("final result: Are you sure? (Remapped by outer)")
        }
    }
```
